Introduce RestrictUsersAdmission plugin for collaboration

Miciah opened the pull request:
Add the `RestrictUsersAdmission` admission plugin for collaboration.

---

openshift-bot, please [test]!

Miciah commented:
Looks like flake #10988.  Please [test] again!

Miciah commented:
Looks like flake #11649.  Please [test] again!

Miciah commented:
This PR implements proposal #9932 to restrict role bindings based on project annotations.  The implementation is based on deads's WIP PR #9885.  @deads2k, @smarterclayton, do you have time to review the approach in this PR?

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,353 @@
+const (
+	// CollaboratorAnnotation is the key for an annotation on Namespace that
+	// specifies collaborators on the namespace.
+	CollaboratorAnnotation = "authorization.openshift.io/collaborators"
```
In 3.3, the kube infrastructure to register different groups was hopelessly broken.  In 3.4, it's only mostly broken.  In 3.5, it mostly works.

I think I'd prefer to see this as a separate API group and resource.  That would allow "clean" management via `oc` and all of our usual tooling.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,353 @@
+
+	// Type is the type of collaborators against which this CollaboratorMatcher
+	// may be matched.
+	Type CollaboratorType `json:"type"`
```
Wrap these and call them `Spec`. I can envision a world where this controller driven and a controller drives updates to status to describe matches.  That would prevent multiple lookup/matching calls on every API server and allow an admin to see who actually matches for something like content assist from a `Status` field.

deads2k replied:
```go
// only one may be non-nil
type CollaboratorMatcherSpec struct{
    UserMatcher *UserMatcher
    GroupMatcher *GroupMatcher
    ServiceAccountMatcher *ServiceAccountMatcher
}

type UserMatcher struct{
    Users []string
    UsersInTheseGroups []string
    UsersMatchingTheseSelectors []*unversioned.LabelSelector
}


type GroupMatcher struct{
    Groups []string
    GroupsMatchingTheseSelectors []*unversioned.LabelSelector
}


type ServiceAccountMatcher struct{
    ServiceAccounts []string
    ServiceAccountsInTheseNamespaces []string
}

```

smarterclayton commented:
The previous proposal never addressed the question of how an end user
would get access to add other people to collaborate in a system where
this was primarily to prevent abuse.  How would this be used to allow
for an online-like cluster (where collaboration was reasonable for
arbitrary user emails)?

Does this presuppose a higher level system for that?

I'd like to be able to answer that question before we add that code.

deads2k commented:
> Does this presuppose a higher level system for that?

Yes.  This assumes an external workflow could be built to handle workflow problems like "apply, approve" and "invite, accept" flows.

Miciah commented on api/protobuf-spec/github_com_openshift_origin_pkg_security_api_v1.proto:
```diff
@@ -5,7 +5,7 @@ syntax = 'proto2';
 
 package github.com.openshift.origin.pkg.security.api.v1;
 
-import "github.com/openshift/origin/pkg/quota/api/v1/generated.proto";
```
I don't know why this is getting deleted.

Miciah commented on pkg/security/api/v1/generated.proto:
```diff
@@ -5,7 +5,7 @@ syntax = 'proto2';
 
 package github.com.openshift.origin.pkg.security.api.v1;
 
-import "github.com/openshift/origin/pkg/quota/api/v1/generated.proto";
```
I don't know why this is getting deleted.

Miciah commented:
@deads2k, could you give this PR another look?

deads2k commented on pkg/authorization/admission/restrictusers/api/types.go:
```diff
@@ -0,0 +1,59 @@
+// a string name on the name of a group to which the user belongs, or
+// a label selector.
+type UserMatcher struct {
+	Users                       []string
```
godoc all API fields.
I'm assuming these are literal string matches.

deads2k commented on pkg/authorization/admission/restrictusers/api/types.go:
```diff
@@ -0,0 +1,59 @@
+type UserMatcher struct {
+	Users                       []string
+	UsersInTheseGroups          []string
+	UsersMatchingTheseSelectors []*unversioned.LabelSelector
```
I think `Selector` will do.

deads2k commented on pkg/authorization/admission/restrictusers/api/types.go:
```diff
@@ -0,0 +1,59 @@
+// a label selector.
+type UserMatcher struct {
+	Users                       []string
+	UsersInTheseGroups          []string
```
`Groups` with good docs.

deads2k commented on pkg/authorization/admission/restrictusers/api/types.go:
```diff
@@ -0,0 +1,59 @@
+// the service-account name or the name of the service account's namespace.
+type ServiceAccountMatcher struct {
+	ServiceAccounts                  []string
+	ServiceAccountsInTheseNamespaces []string
```
`Namespaces`?

deads2k commented on pkg/authorization/admission/restrictusers/api/types.go:
```diff
@@ -0,0 +1,59 @@
+// ServiceAccountMatcher matches a service account by a string match on either
+// the service-account name or the name of the service account's namespace.
+type ServiceAccountMatcher struct {
+	ServiceAccounts                  []string
```
This needs to a slice of structs which have `namespace` and `name`.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,279 @@
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
+	policy := authorizationapi.GetPolicyBindingName(binding.RoleRef.Namespace)
+	policybinding, _ := q.policyBindingLister.PolicyBindings(ns).Get(policy)
```
The old object is provided via `a.GetOldObject()`.  Use that instead of a cached. copy.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,279 @@
+	for _, matcher := range matchersList.Items {
+		// TODO: Use a cache for users, groups, and serviceaccounts.
+		switch {
+		case matcher.Spec.UserMatcher != nil:
```
I'd like to see an interface defined that checks a subject.  Something like `Allowed(subject) error`.  I would be able to take a `collaboratormatcherspec` and get an allowed back out and then something that provides a union of `subjectcheckers`.  It'll make it easier to read.

Miciah replied:
Are you suggesting something like the following?

```go
func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
  // ...
  allowed := NewEmptyAllowed()
  allowed.SetNamespace(ns)
  allowed.SetClients(q.kclient, q.oclient)

  for _, matcher := range matchersList.Items {
    // NewAllowed(nil) is the same as NewEmptyAllowed().
    // allowed.Merge(NewEmptyAllowed()) is a no-op.
    allowed = allowed.Merge(NewAllowed(matcher.UserMatcher))
    allowed = allowed.Merge(NewAllowed(matcher.GroupMatcher))
    allowed = allowed.Merge(NewAllowed(matcher.ServiceAccountMatcher))
  }
  // ...
}
```

For comparison, here is what I have now:

```go
func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
  // ...
  checker, err := NewRoleBindingRestrictions(ns, q.kclient, q.oclient)
  if err != nil {
    return admission.NewForbidden(a, err)
  }

  for _, matcher := range matchersList.Items {
    checker.AddMatcher(&matcher)
  }
  // ...
}
```

The latter seems simpler to me, but maybe I misunderstand the suggestion.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,279 @@
+			}
+
+			for _, literal := range userInGroupLiterals {
+				for _, group := range user.Groups {
```
There should be a groupcache which the API server already maintains to get group information for a user.  Use that instead of hitting the API for it.

deads2k replied:
Oh, I see.  You need labels too.  nevermind.

deads2k commented:
Let's the types well documented and then make the `SubjectChecker` (or a better name if you have it) build for each kind of matcher.  That's going to be the easiest thing for api-review to look at to understand behavior.

Also, `RoleBindingRestrictions` as the overall type?

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,279 @@
+		return nil
+	}
+
+	matchersList, err := q.oclient.CollaboratorMatchers(ns).List(kapi.ListOptions{})
```
If I have no matchers, seems like I should let anyone bind in.  That would be backwards compatible.

Miciah commented:
I have updated the PR:

• Renamed fields in `UserMatcher`, `GroupMatcher`, and `ServiceAccountMatcher` to be more succinct.
• Made `ServiceAccountMatcher` into a tuple of serviceaccount name and namespace.
• Copied docstrings from `v1/api/types.go` into `api/types.go`.
• Merged `pkg/authorization/admission/restrictusers/api/` into `pkg/authorization/api/`.
• Moved `pkg/authorization/admission/restrictusers/registry/collaboratormatcherregistry` under `pkg/authorization/registry/`.
• Changed `func (r *restrictUsersAdmission) Admit` to use `a.GetOldObject()` and deleted all obviated informer code.
• Defined a `SubjectChecker` interface with `AddMatcher` and `Allowed` methods.
• Defined a `RoleBindingRestrictions` type that implements `SubjectChecker`, and changed `RestrictUsersAdmission` to use it.
• Updated tests, added some docstrings, probably made some other minor changes.

@deads2k, what do you think now?

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,181 @@
+// RoleBindingRestrictions aggregates rules specified by CollaboratorMatcher
+// objects in a namespace and uses these rules to determine whether subjects
+// (users, groups, or service accounts) are collaborators on the namespace.
+type RoleBindingRestrictions struct {
```
I'd like one struct for each kind of `Matcher`.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,181 @@
+// the namespace.
+func (r *RoleBindingRestrictions) Allowed(subject *kapi.ObjectReference) error {
+	switch subject.Kind {
+	case authorizationapi.SystemUserKind:
```
I'd like to start by managing these inside their respect user and group restrictions, but match them based only on literals.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,181 @@
+// based on a set of CollaboratorMatcher objects, which must be added to the
+// SubjectChecker before use.
+type SubjectChecker interface {
+	Allowed(kapi.ObjectReference) error
```
Looking at it in action, I think we wnt to

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,181 @@
+func (r *RoleBindingRestrictions) AddMatcher(cm *authorizationapi.CollaboratorMatcher) error {
+	switch {
+	case cm.Spec.UserMatcher != nil:
+		r.userLiterals = append(r.userLiterals, cm.Spec.UserMatcher.Users...)
```
This is exactly the kind of code I want to avoid.  Building up a structure like this adds a layer that I don't think we need.

deads2k commented on pkg/authorization/api/types.go:
```diff
@@ -434,3 +434,86 @@ type ClusterRoleList struct {
+// the namespace to which the CollaboratorMatcher belongs.  If any one of those
+// CollaboratorMatcher objects matches a subject, that subject is considered to
+// be a collaborator on the namespace.
+type CollaboratorMatcher struct {
```
`RoleBindingRestriction`

deads2k commented on pkg/authorization/api/types.go:
```diff
@@ -434,3 +434,86 @@ type ClusterRoleList struct {
+	// Users specifies a list of literal user names.
+	Users []string
+
+	// Groups specifies a list of literal group names.
```
`Groups is a list of groups used to match against an individual user's groups.  If the user is a member of one of the whitelisted groups, he's allowed to be bound to a role.`

deads2k commented on pkg/authorization/api/types.go:
```diff
@@ -434,3 +434,86 @@ type ClusterRoleList struct {
+	// ServiceAccounts specifies a list of literal service-account names.
+	ServiceAccounts []*ServiceAccountInNamespace
+
+	// Namespaces specifies a list of literal namespace names.
```
add, `ServiceAccounts from inside the whitelisted namespaces are allowed to be bound to roles.`

deads2k commented on pkg/authorization/api/types.go:
```diff
@@ -434,3 +434,86 @@ type ClusterRoleList struct {
+
+// ServiceAccountInNamespace specifies a service account and namespace by their
+// names.
+type ServiceAccountInNamespace struct {
```
`ServiceAccountReference`

deads2k commented on pkg/authorization/api/util.go:
```diff
@@ -0,0 +1,15 @@
+	case collaboratorMatcher.Spec.GroupMatcher != nil:
+		return "Group"
+	case collaboratorMatcher.Spec.ServiceAccountMatcher != nil:
+		return "Service Account"
```
No space to match CLI commands.

deads2k replied:
If this is only used for printer and describer, move it to the CLI code.

deads2k commented on pkg/authorization/registry/collaboratormatcherregistry/etcd.go:
```diff
@@ -0,0 +1,52 @@
+	store := &registry.Store{
+		NewFunc:           func() runtime.Object { return &authorizationapi.CollaboratorMatcher{} },
+		NewListFunc:       func() runtime.Object { return &authorizationapi.CollaboratorMatcherList{} },
+		QualifiedResource: authorizationapi.Resource("collaboratormatchers"),
```
rolebindingrestriction

deads2k commented on pkg/authorization/registry/collaboratormatcherregistry/strategy.go:
```diff
@@ -0,0 +1,74 @@
+var Strategy = strategy{kapi.Scheme}
+
+func (strategy) NamespaceScoped() bool {
+	return false
```
`true`

deads2k commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -1060,3 +1065,70 @@ func printAppliedClusterResourceQuotaList(list *quotaapi.AppliedClusterResourceQ
+		if len(cm.Spec.UserMatcher.Users) != 0 {
+			matchType = "literal"
+			matcher = strings.Join(cm.Spec.UserMatcher.Users, ", ")
+		} else if len(cm.Spec.UserMatcher.Groups) != 0 {
```
This isn't exclusive.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,158 @@
+// collaborators.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create),
```
You need to handle updates too.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,158 @@
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
```
Break out this chunk into a method please.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,158 @@
+		for _, newSubject := range binding.Subjects {
+			foundExisting := false
+			for _, existingSubject := range oldbinding.Subjects {
+				if existingSubject.Kind == newSubject.Kind &&
```
ought to be able to just check straight `==` right?  We'll never false negative and false positive is ok.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,158 @@
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
```
please remove space.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,158 @@
+		checker.AddMatcher(&matcher)
+	}
+
+	// TODO: Aggregate errors and return an error that lists all subjects that are
```
I see this TODO as required for merge.  Otherwise people will hate our API server even more.

deads2k commented:
We'll finish the API review first, but you'll need an integration test demonstrating success and failure for both create and update of rolebindings.

Miciah commented:
Updated:

* Added `admission.Update` to the list of operations that `RestrictUsersAdmission` handles.
* Factored `func rolebindingsDelta` out of `func (q *restrictUsersAdmission) Admit`.
* Changed `func (q *restrictUsersAdmission) Admit` to build and return a list of all rejected subjects instead of returning an error with just the first rejected subject.
* Deleted `AddMatcher` from the `SubjectChecker` interface.
* Moved `CollaboratorMatcherType` from `pkg/authorization/api/util.go` to `pkg/cmd/cli/describe/helpers.go`.
* Renamed a lot of things:
  * `CollaboratorMatcher` to `RoleBindingRestriction` (and likewise for related types, methods, directories, and files),
  * `UserMatcher` to `UserRestriction`,
  * `GroupMatcher` to `GroupRestriction`,
  * `ServiceAccountMatcher` to `ServiceAccountRestriction`,
  * `ServiceAccountInNamespace` to `ServiceAccountReference`.
  * `func RoleBindingRestrictions` to `func aggregatedRoleBindingRestrictions`,
* Improved docstrings for `UserRestriction` and `ServiceAccountRestriction`.
* Changed `func aggregatedRoleBindingRestrictions` to store lists of `UserRestriction`, `GroupRestriction`, and `ServiceAccountRestriction` instead of lists of literals and selectors—@deads2k, is this what you meant, or is it your intention to make `RoleBindingRestrictionSpec` itself a `SubjectChecker`? In the latter case, we will need to revise the interface to allow the required context (namespace and either client objects or user, group, and serviceaccount resources) to be passed along.  This would also imply that we would have non-API `RoleBindingRestrictionSpec` objects with more than one of `UserRestriction`, `GroupRestriction`, and `ServiceAccountRestriction` non-nil, which seems overloaded and potentially confusing.
* Changed `func (r *aggregatedRoleBindingRestrictions) Allowed` to allow system users for which some `UserRestriction` matches on user name and system groups for which some `GroupRestriction` matches on group name.
* Added `roleBindingRestrictionContext` to store namespace and client objects, and embed it in `aggregatedRoleBindingRestrictions`.
* Factored `func userAllowed`, `func groupAllowed`, and `func serviceaccountAllowed` out of `func (r *aggregatedRoleBindingRestrictions) Allowed`.
* Added `func ValidateRoleBindingRestrictionUser`, `func ValidateRoleBindingRestrictionGroup`, and `func ValidateRoleBindingRestrictionServiceAccount`, which validate that rolebinding restriction specs are nonempty and have no invalid selectors.
* Changed `rolebindingrestrictionregistry` to use `SimpleNameGenerator`.
* Changed `func rolebindingrestrictionregistry.NamespaceScoped` to `true`.
* Added `func (obj *RoleBindingRestrictionList) GetObjectKind` and `func (obj *RoleBindingRestriction) GetObjectKind` to `pkg/authorization/api/register.go` and likewise to `pkg/authorization/api/v1/register.go`—not sure whether this is correct.
* Fixed `printRoleBindingRestriction` to print subjects from all match types (e.g., literals, groups, and selectors for users) specified in a rolebinding restriction spec, and cap the total number of subjects printed for a given rolebinding restriction to 3.
* Moved `pkg/authorization/registry/rolebindingrestrictionregistry/` to `pkg/authorization/registry/rolebindingrestriction/` and fixed the package declaration.

However, I do not seem to be correctly registering the API—`oc get resourceaccessreview` reports that "the server does not allow this method on the requested resource".  @deads2k, do you see what the problem is?

deads2k commented:
> However, I do not seem to be correctly registering the API—oc get resourceaccessreview reports that "the server does not allow this method on the requested resource". @deads2k, do you see what the problem is?

`resourceaccessreview` can only be created.

Miciah commented:
Sorry, I meant `oc get rolebindingrestrictions` gives that error message.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,168 @@
+			}
+		}
+		result = append(result, y)
+	nexty:
```
no labels please.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,168 @@
+// each subject in the binding must be matched by some rolebinding restriction
+// in the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
```
you need to catch policybindings too.  You should coerce the policybindings down to the different rolebindings and check them each individually.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,247 @@
+
+// SubjectChecker determines whether rolebindings on a subject (user, group, or
+// service account) are allowed in a namespace.
+type SubjectChecker interface {
```
There should be a struct for each matcher type that fulfills this interface.  I'm ok also passing through a context (struct like you have below), but I still can't reasonably read through what this is trying to do with its information.

deads2k replied:
I'm expecting unit tests for individual subject checkers, then unit tests for the admission plugin focus on "did I describe the deltas correctly and run all the checks in the chain"

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,247 @@
+		}
+
+	case authorizationapi.UserKind:
+		user, err := r.ctx.oclient.Users().Get(subject.Name)
```
Try to match literals before checking for label matches.  We want to allow pre-binding if possible.  Similarly, for user matching, try to use the userToGroup cache to pull the group information without requiring a user be present ahead of time.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,247 @@
+		}
+
+	case authorizationapi.ServiceAccountKind:
+		sa, err := r.ctx.kclient.Core().ServiceAccounts(r.ctx.namespace).Get(subject.Name)
```
Why do you need the actual SA?  Also, allow the restriction to match with an empty namespace if the SA and restriction are in the same namespace.  This matches other reference behavior.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,168 @@
+		return nil
+	}
+
+	rbrCtx, err := newRoleBindingRestrictionContext(ns, q.kclient, q.oclient)
```
At this point, build the slice of `SubjectCheckers`, then construct the `UnionSubjectChecker` (I think that's just `type UnionSubjectChecker []SubjectCheckers`, then run the check with the context.

deads2k commented on pkg/authorization/api/types.go:
```diff
@@ -434,3 +434,90 @@ type ClusterRoleList struct {
+	// Name is the name of the service account.
+	Name string
+
+	// Namespace is the name of the namespace.
```
If the namespace is empty, then the reference means "SA with this name in the current namespace"

deads2k commented on pkg/authorization/registry/rolebindingrestriction/etcd.go:
```diff
@@ -0,0 +1,52 @@
+	store := &registry.Store{
+		NewFunc:           func() runtime.Object { return &authorizationapi.RoleBindingRestriction{} },
+		NewListFunc:       func() runtime.Object { return &authorizationapi.RoleBindingRestrictionList{} },
+		QualifiedResource: authorizationapi.Resource("rolebindingrestrictions"),
```
singular.

deads2k commented on pkg/cmd/server/start/admission.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"k8s.io/kubernetes/pkg/util/sets"
 
 	// Admission control plug-ins used by OpenShift
+	_ "github.com/openshift/origin/pkg/authorization/admission/restrictusers"
```
Set yourself to default off.

deads2k commented:
> RoleBindingRestriction as the name of an object whose existence allows something is a little odd

Its existence is purely restrictive in nature.  "Allow these and nothing else" as opposed to allow everything.

The alternative option is to wire up the admission plugin as default-off, but flipping the switch would be more painful then and the expression of what you want starts requiring things like 'allow all', which isn't particularly useful but could be done.

Also, I think critics are required to provide suggestions. :)

Miciah commented:
Updated:

* Added `WantsGroupCache` interface for plugin for admission plugins, and changed `restrictUsersAdmission` to use it.
* Added `groupCache *usercache.GroupCache` to `restrictUsersAdmission`.
* Renamed `func rolebindingsDelta` to `func objectReferenceDelta`.
* Rewrote `func objectReferenceDelta` without the labelled goto.
* Changed `func (q *restrictUsersAdmission) Admit` to handle policybindings in addition to rolebindings.
* Defined the `UnionSubjectChecker` type as `[]SubjectChecker`, made it a `SubjectChecker`, and replaced `aggregatedRoleBindingRestrictions` with it.
* Added `userToLabelSet` and `userToLabelSet` maps to `RoleBindingRestrictionContext` and defined `func (ctx *RoleBindingRestrictionContext) labelSetForUser` and `func (ctx *RoleBindingRestrictionContext) labelSetForGroup`, which perform API queries as needed to lazily build these maps.
* Added `groupCache *usercache.GroupCache` to `RoleBindingRestrictionContext` and defined `func (ctx *RoleBindingRestrictionContext) groupsForUser`.
* Added a `*RoleBindingRestrictionContext` parameter and `bool` return value to the `Allowed` method of `SubjectChecker`.
* Defined `UserSubjectChecker`, `GroupSubjectChecker`, and `ServiceAccountSubjectChecker`, which all implement `SubjectChecker`, and associated constructors and methods, which use the new `RoleBindingRestrictionContext` methods.
* Expanded the docstring for the `Namespace` field of `ServiceAccountReference`.
* Fixed the etcd name of "rolebindingrestriction" resources (singular).
* Made the "RestrictSubjectBindings" plugin disabled by default.

However, there is now a race condition in the test that involves the group cache, which I am trying to figure out; meanwhile, a 100 millisecond sleep enables the test to pass consistently for me.  Once I figure that out, I plan to split the tests into `subjectchecker_test.go` and `restrictuser_test.go`.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,248 @@
+
+// objectReferenceDelta returns the relative complement of []ObjectReference a
+// in []ObjectReference b (i.e., a∖b).
+func objectReferenceDelta(a, b []kapi.ObjectReference) []kapi.ObjectReference {
```
names.  needles and haystack maybe?  Something better than a and b at any rate.

Miciah replied:
Would "left" and "right" be acceptable? Or "lhs" and "rhs"?

deads2k replied:
> Would "left" and "right" be acceptable? Or "lhs" and "rhs"?

Not really.  This is a test of "set of authority" and "things to check against the set of the authority" as opposed to something like a comparison.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,297 @@
+
+// groupsForUser returns the groups for the given user subject.
+func (ctx *RoleBindingRestrictionContext) groupsForUser(subject kapi.ObjectReference) ([]*userapi.Group, error) {
+	if subject.Kind != authorizationapi.UserKind {
```
When called for a `system:user`, just return an empty set, but not an error.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,297 @@
+
+// labelSetForUser returns the label set for the given user subject.
+func (ctx *RoleBindingRestrictionContext) labelSetForUser(subject kapi.ObjectReference) (labels.Set, error) {
+	if subject.Kind != authorizationapi.UserKind {
```
When called for a `system:user`, just return an empty set, but not an error.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,297 @@
+// UserSubjectChecker determines whether a user subject is allowed in
+// rolebindings in the project.
+type UserSubjectChecker struct {
+	SubjectChecker
```
huh?

deads2k commented:
@openshift/api-review this is ready for a broader API review

deads2k commented:
@smarterclayton

Miciah commented:
@smarterclayton, thanks for the comments! I have updated the comments in `types.go`.  Did your comments constitute an API review, or is that still pending completion?

smarterclayton commented on pkg/authorization/api/v1/types.go:
```diff
@@ -476,3 +476,92 @@ type ClusterRoleList struct {
+// namespace.
+type ServiceAccountRestriction struct {
+	// ServiceAccounts specifies a list of literal service-account names.
+	ServiceAccounts []*ServiceAccountReference `json:"serviceaccounts" protobuf:"bytes,1,rep,name=serviceaccounts"`
```
Why is this a pointer?

Miciah replied:
I'm not sure, but I was probably following the example of the `Selectors []*unversioned.LabelSelector` fields in `UserRestriction` and `GroupRestriction`, so I will change this if there is no good reason to use pointers for those fields.

smarterclayton commented on pkg/authorization/api/v1/types.go:
```diff
@@ -476,3 +476,92 @@ type ClusterRoleList struct {
+	Groups []string `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+
+	// Selectors specifies a list of label selectors over group labels.
+	Selectors []*unversioned.LabelSelector `json:"labels" protobuf:"bytes,2,rep,name=labels"`
```
Why is this a pointer?

Miciah replied:
Based on deads's suggested data structure: https://github.com/openshift/origin/pull/11743#discussion_r86876201.  @deads2k, any reason to use pointers?

smarterclayton commented on pkg/authorization/api/v1/types.go:
```diff
@@ -476,3 +476,92 @@ type ClusterRoleList struct {
+	Groups []string `json:"groups" protobuf:"bytes,2,rep,name=groups"`
+
+	// Selectors specifies a list of label selectors over user labels.
+	Selectors []*unversioned.LabelSelector `json:"labels" protobuf:"bytes,3,rep,name=labels"`
```
Pointer?

smarterclayton commented:
With that change, api-approved.

On Tue, Dec 6, 2016 at 10:10 AM, David Eads <notifications@github.com>
wrote:

> *@deads2k* commented on this pull request.
> ------------------------------
>
> In pkg/authorization/api/v1/types.go
> <https://github.com/openshift/origin/pull/11743>:
>
> > +	Groups []string `json:"groups" protobuf:"bytes,2,rep,name=groups"`
> +
> +	// Selectors specifies a list of label selectors over user labels.
> +	Selectors []*unversioned.LabelSelector `json:"labels" protobuf:"bytes,3,rep,name=labels"`
> +}
> +
> +// GroupRestriction matches a group either by a string match on the group name
> +// or a label selector applied to group labels.
> +type GroupRestriction struct {
> +	// Groups is a list of groups used to match against an individual user's
> +	// groups. If the user is a member of one of the whitelisted groups, the user
> +	// is allowed to be bound to a role.
> +	Groups []string `json:"groups" protobuf:"bytes,1,rep,name=groups"`
> +
> +	// Selectors specifies a list of label selectors over group labels.
> +	Selectors []*unversioned.LabelSelector `json:"labels" protobuf:"bytes,2,rep,name=labels"`
>
> Based on deads's suggested data structure: #11743
> <https://github.com/openshift/origin/pull/11743> (comment). @deads2k
> <https://github.com/deads2k>, any reason to use pointers?
>
> Just copy/pasta.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11743>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-stp8hRr0xeUpePdzbXtJUDKbkPks5rFXrNgaJpZM4Kn2FF>
> .
>

Miciah commented:
Updates:

* Fixed `ValidateRoleBindingRestriction` to expect `Namespace` in `RoleBindingRestriction`'s metadata.
* Changed "rolebindingrestriction" to "rolebindingrestrictions" in `etcd.go`.
* Improved formatting in the printer and describer code.
* Factored `func NewSubjectChecker(spec *authorizationapi.RoleBindingRestrictionSpec) (SubjectChecker, error)` out of `func (q *restrictUsersAdmission) Admit`.
* Added "RestrictSubjectBindings" to `openshiftAdmissionControlPlugins` and `CombinedAdmissionControlPlugins` in `pkg/cmd/server/origin/master_config.go`.
* Added integration tests.
* Split unit tests into `groupcache_test.go`, `subjectchecker_test.go`, and `restrictusers_test.go`.
* Added unit tests for policybindings (coverage is now 80.8%).

Questions:

* Is the 100ms sleep in the test code to allow the group cache to synch acceptable?
* Should I add `omitempty` to the fields of `UserRestriction`, `GroupRestriction`, and `ServiceAccountRestriction` and to the eponymous fields of `RoleBindingRestrictionSpec`?
* What else is needed to get this merged?

Miciah commented:
Looks like flake #11353.  Please [test] again!

Miciah commented:
Whoops, missed this one—I'll update shortly:
```
--- FAIL: TestAdmissionPluginNames (0.00s)
	admissionconfig_test.go:93: openshift admission plugins must be prefixed with openshift.io/ RestrictSubjectBindings
```

Miciah commented:
Updated to fix the following failures:
```
--- FAIL: TestAdmissionPluginNames (0.00s)
	admissionconfig_test.go:93: openshift admission plugins must be prefixed with openshift.io/ RestrictSubjectBindings
```
and
```
--- FAIL: TestKubeAdmissionControllerUsage (0.00s)
	admission_sync_test.go:57: RestrictSubjectBindings not used
```

Miciah commented:
Added "rolebindingrestrictions" to `pkg/cmd/server/bootstrappolicy/policy.go` to fix this failure:
```
--- FAIL: TestClusterReaderCoverage (4.17s)
	authorization_test.go:171: cluster-reader role is missing map[{ rolebindingrestrictions}:true].  Check pkg/cmd/server/bootstrappolicy/policy.go.
```

Miciah commented:
In case we're trying to keep track, I also saw what looks like flake #9681 in the last test run:
```
deploymentconfigs
/data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:845
  when run iteratively [Conformance]
  /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:196
    should immediately start a new deployment [It]
    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:195

    Expected error:
        <*errors.errorString | 0xc820184aa0>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    not to have occurred

    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:194
```
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/9398/consoleFull

Miciah commented:
Update with regenerated `test/testdata/bootstrappolicy/bootstrap_cluster_roles.yaml`, which needed to be regenerated after the last update to `pkg/cmd/server/bootstrappolicy/policy.go`.

deads2k commented:
> Will this type be a blocker to moving to upstream RBAC until upstream has parity?

Given initializers and external update validation admission, it doesn't have to be.

abhgupta commented:
@deads2k looking for a final LGTM - is this ready for a merge now?

deads2k commented:
> @deads2k looking for a final LGTM - is this ready for a merge now?

You can't have a sleep in your test. Check the condition you need.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker_test.go:
```diff
@@ -0,0 +1,273 @@
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
```
> @deads2k, I've added a condition check for the group cache, but AFAICT a sleep is necessary in order to relinquish execution and allow the goroutine that initialises the cache a chance to run, so I've kept it (but reduced it to 1ms, in a for loop). Is that acceptable?

The condition check makes its ok.  10 milliseconds would be a better wait.

deads2k replied:
so the sleep is good enough for now, but what you probably really wanted was to create an interface for the groupcache and implement your own mock for the test.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,232 @@
+// each subject in the binding must be matched by some rolebinding restriction
+// in the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if !q.groupCache.Running() {
```
move this check after check of the resource so that you don't fail requests you don't have to.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,232 @@
+	var subjects, oldSubjects []kapi.ObjectReference
+
+	obj, oldObj := a.GetObject(), a.GetOldObject()
+	switch a.GetResource().GroupResource() {
```
if you aren't one of these resources, return early.

deads2k commented on pkg/authorization/admission/restrictusers/restrictusers.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// TODO: Cache rolebinding restrictions.
+	roleBindingRestrictionList, err := q.oclient.RoleBindingRestrictions(ns).
```
Open an issue for 3.5 and be sure to fix it before release.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,319 @@
+	for _, checker := range []SubjectChecker(checkers) {
+		allowed, err := checker.Allowed(subject, ctx)
+		if err != nil {
+			return false, err
```
Track errors and report them as an aggregate at the end.  I think its reasonable to keep checking other `SubjectCheckers` in the case of error.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,319 @@
+
+// Allowed determines whether the given subject is allowed in rolebindings in
+// the project.
+func (checkers UnionSubjectChecker) Allowed(subject kapi.ObjectReference,
```
nit: no line break please.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker.go:
```diff
@@ -0,0 +1,319 @@
+type UnionSubjectChecker []SubjectChecker
+
+// NewUnionSubjectChecker returns a new UnionSubjectChecker.
+func NewUnionSubjectChecker(checkers []SubjectChecker) (UnionSubjectChecker, error) {
```
doesn't look like you ever return an error.

deads2k commented on pkg/authorization/admission/restrictusers/subjectchecker_test.go:
```diff
@@ -0,0 +1,273 @@
+	var (
+		userBobRef = kapi.ObjectReference{
+			Kind:      authorizationapi.UserKind,
+			Namespace: "namespace",
```
users don't have namespaces.

deads2k commented on pkg/authorization/api/v1/register.go:
```diff
@@ -45,6 +46,13 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 	return nil
 }
+
+func (obj *RoleBindingRestrictionList) GetObjectKind() unversioned.ObjectKind { return &obj.TypeMeta }
```
I expected you to get these automatically.  Can you delete them and have it still compile/work?

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -137,7 +137,7 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 
 				authorizationapi.NewRule(read...).Groups(authzGroup).Resources("clusterpolicies", "clusterpolicybindings", "clusterroles", "clusterrolebindings",
-					"policies", "policybindings", "roles", "rolebindings").RuleOrDie(),
+					"policies", "policybindings", "roles", "rolebindings", "rolebindingrestrictions").RuleOrDie(),
```
a project admin should be allowed to see the `rolebindingrestrictions`

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -330,6 +331,7 @@ var (
 	// openshiftAdmissionControlPlugins gives the in-order default admission chain for openshift resources.
 	openshiftAdmissionControlPlugins = []string{
 		"ProjectRequestLimit",
+		"openshift.io/RestrictSubjectBindings",
```
run after originnamespacelifecycle

deads2k commented:
Comments in a separate commit if you would.  Almost done.

Miciah commented:
I've updated generated files and pushed a new commit addressing the latest review comments.

Miciah commented:
Saw what could be flake #9681 again:
```
deploymentconfigs
/data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:845
  with multiple image change triggers [Conformance]
  /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:335
    should run a successful deployment with multiple triggers [It]
    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:326

    Expected error:
        <*errors.errorString | 0xc8200d3780>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    not to have occurred

    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:325
```
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/9550/consoleFull

Miciah commented:
Flake https://github.com/openshift/origin/issues/12235 (https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/9552/consoleFull).

New flake https://github.com/openshift/origin/issues/12251 (https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_networking/2380/consoleFull).

Also, fixed the bootstrap policy again.  We want to allow `admin` *and* `cluster-reader` to be able to read `rolebindingrestrictions`, right? With only the former, I get the following error:
```
--- FAIL: TestClusterReaderCoverage (4.09s)
	authorization_test.go:171: cluster-reader role is missing map[{ rolebindingrestrictions}:true].  Check pkg/cmd/server/bootstrappolicy/policy.go.
```

deads2k commented:
Two minor comments I'd like to see fixed before merge.  Go ahead and squash them in and this will be ready.

Miciah commented:
I've made those two changes:
https://github.com/openshift/origin/pull/11743/files#diff-93fae1b66fbece23edf504e5e5aebf1aR42
https://github.com/openshift/origin/pull/11743/files#diff-cbe61dec94c120466223271f21bcd664R88
and squashed.

deads2k commented:
lgtm [merge]

deads2k commented:
Disabled the merge tag. 

@ncdc Are we good to merge this or are you close enough we should hold it?

ncdc commented:
I am probably still a few days away.

deads2k commented:
[merge]

Miciah commented:
Flake https://github.com/openshift/origin/issues/11016.
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/9597/consoleFull
```
[networking][router] openshift routers
/data/src/github.com/openshift/origin/test/extended/router/scoped.go:135
  The HAProxy router
  /data/src/github.com/openshift/origin/test/extended/router/scoped.go:134
    should override the route host with a custom value [It]
    /data/src/github.com/openshift/origin/test/extended/router/scoped.go:133

    Expected error:
        <*errors.errorString | 0xc8200f5600>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    not to have occurred

    /data/src/github.com/openshift/origin/test/extended/router/scoped.go:95
```
